tests/test_integration.py

```diff
@@ -749,5 +768,149 @@ def test_train_single_task_regression_reaction(self,
 
+    @parameterized.expand([
+        (
+                'chemprop',
+                'chemprop',
+                'auc',
+                0.699453,
+                ['--number_of_molecules', '2', '--data_path', os.path.join(TEST_DATA_DIR, 'classification_multimolecule.csv')]
+        )
+    ])
+    def test_single_task_multimolecule_classification(self,
+                                          name: str,
+                                          model_type: str,
+                                          metric: str,
+                                          expected_score: float,
+                                          train_flags: List[str] = None):
+        with TemporaryDirectory() as save_dir:
+            # Train
+            self.train(
+                dataset_type='classification',
+                metric=metric,
+                save_dir=save_dir,
+                model_type=model_type,
+                flags=train_flags
+            )
+
+            # Check results
+            test_scores_data = pd.read_csv(os.path.join(save_dir, TEST_SCORES_FILE_NAME))
+            test_scores = test_scores_data[f'Mean {metric}']
+
+            mean_score = test_scores.mean()
+            self.assertAlmostEqual(mean_score, expected_score, delta=DELTA * expected_score)
+
+    @parameterized.expand([
+        (
+                'chemprop',
+                'chemprop',
+                ['--fingerprint_type', 'MPN'],
+        ),
+        (
+                'chemprop',
+                'chemprop',
+                ['--fingerprint_type', 'last_FFN']
+        )
+    ])
+    def test_single_task_fingerprint(self,
+                                            name: str,
+                                            model_type: str,
+                                            fingerprint_flags: List[str],
+                                            train_flags: List[str] = None,
+                                     ):
+        with TemporaryDirectory() as save_dir:
+            # Train
+            dataset_type = 'classification'
+            self.train(
+                dataset_type=dataset_type,
+                metric='auc',
+                save_dir=save_dir,
+                model_type=model_type,
+                flags=train_flags
+            )
+
+            # Predict
+            fingerprint_path = os.path.join(save_dir, 'fingerprints.csv')
+            self.fingerprint(
+                dataset_type=dataset_type,
+                checkpoint_dir=save_dir,
+                fingerprint_path=fingerprint_path,
+                fingerprint_flags=fingerprint_flags
+            )
+
+            # Check to make sure that fingerprints are generated for all input molecules
+            fingerprints = pd.read_csv(fingerprint_path)
+            test_input = pd.read_csv(os.path.join(TEST_DATA_DIR, f'{dataset_type}_test_smiles.csv'))
+
+            self.assertEqual(list(fingerprints['smiles'].values), list(test_input['smiles'].values))
```

Comment:
Could use a test where we see if a set of molecules actually returns the fingerprint values we expect. Or match a hash or aggregation of the fingerprint values (sum or something).

Reply:
That's what I originally wanted to do but I wasn't sure how different the fingerprints would be run-to-run and if that's something we could feasibly get around.

A hash might be too sensitive to the nondeterminism of the fingerprint outputs if we're training a new network each run, but I like the idea of summing the fingerprint values. I'll have to see how different the fingerprints end up. I'll work on that.

Reply:
This works for comparing the sums of the columns:

`           test_fingerprint_sums = fingerprints.sum()
            true_fingerprint_sums = true_fingerprints.sum()

            similar_fingerprint_count = 0

            for test, true in zip(test_fingerprint_sums, true_fingerprint_sums):
                if abs(test - true) <= true * DELTA:
                    similar_fingerprint_count += 1

            self.assertTrue(similar_fingerprint_count/len(fingerprints.columns) >= 0.95)`

Because a few columns can differ drastically run-to-run, I just check to make sure that the majority are consistent.

Fwiw, checking without summing the columns runs in pretty much the same time:

`            similar_fingerprint_count = 0

            for column in fingerprints.columns:
                for a, b in zip(fingerprints[column].values, true_fingerprints[column].values):
                    if abs(a - b) <= b * DELTA:
                        similar_fingerprint_count += 1

            self.assertTrue(similar_fingerprint_count/fingerprints.size >= 0.95)`

Reply:
Sorry for the delay looking at this. This is a little concerning. They should be nearly deterministic, I wouldn't expect them to vary detectably between runs. Maybe with the exception of some values very close to zero.

I'm going to do a little bit of testing and see if I can nail down where the variation is coming from.

Reply:
I did a test outside of pytest and it showed no variation out past 10 decimal points. Maybe there's some variability that is coming from pytest. I'll try that next.

Reply:
No worries! It's a little dumb in hindsight but I hadn't realized that each training loop uses the same seed for initial weights and biases, hence why I thought there would be more non-determinism in the outputs. 

Double checked the fingerprint outputs and it looks like they are identical. 

```
for column in fingerprints.columns:
    for a, b in zip(fingerprints[column].values, true_fingerprints[column].values):
         self.assertEqual(a, b)
```

Now passes in the tests. I'm not sure why this wasn't passing for me earlier.

Reply:
Okay, just re-ran some tests and

```
for column in fingerprints.columns:
    for a, b in zip(fingerprints[column].values, true_fingerprints[column].values):
        self.assertAlmostEqual(a, b)
```

is in fact working fine. Like I said above, I'm not sure why this wasn't working for me earlier but I don't think it's an issue now.

Reply:
Apologies for the delay, were you able to look into this at all? I'm still seeing some runs pass and others fail (even using a delta) due to variability in the fingerprint output. There will always be some degree of nondeterminism with pytorch (https://pytorch.org/docs/stable/notes/randomness.html). Perhaps comparisons of the latent states are just more sensitive to this nondeterminism than comparing AUC scores as is done in the other tests?

Reply:
Sorry for the delay on my part, I've been held up from doing a final review.

I'm still concerned about the variability that you see. Pytorch acknowledges that it's not strictly controlled, but I haven't seen any variation when I run tests. Are you using a gpu for your local testing?

Maybe it's the aggregation that makes things more sensitive. Like the great magnitude of the values cancel because they're alternating signs and we're just left with remainders dominated by otherwise small variation.

Reply:
Np, I'm still confused by it too. No GPU, I've just been running the tests on my MacBook. 

Here's the full code of the test_single_task_fingerprint test that I've been using:

```
        with TemporaryDirectory() as save_dir:
            # Train
            dataset_type = 'classification'
            self.train(
                dataset_type=dataset_type,
                metric='auc',
                save_dir=save_dir,
                model_type=model_type,
                flags=train_flags
            )

            # Fingerprint
            fingerprint_path = os.path.join(save_dir, 'fingerprints.csv')
            self.fingerprint(
                dataset_type=dataset_type,
                checkpoint_dir=save_dir,
                fingerprint_path=fingerprint_path,
                fingerprint_flags=fingerprint_flags
            )

            fingerprints = pd.read_csv(fingerprint_path).drop(["smiles"], axis=1)
            true_fingerprints = pd.read_csv(os.path.join(TEST_DATA_DIR, f'fp_{fingerprint_flags[1]}_test_true.csv')).drop(["smiles"], axis=1)

            self.assertEqual(fingerprints.size, true_fingerprints.size)

            for column in fingerprints.columns:
                for a, b in zip(fingerprints[column].values, true_fingerprints[column].values):
                    self.assertAlmostEqual(a, b, delta=DELTA*b)
```

Anecdotally, about one out of every 5 or 10 tests fails. Can you see if you can get the test code above to fail after repeat runs?

Reply:
I loaded up your branch to test it out but oddly didn't seem like the true fingerprint file was in the tests/data directory so I couldn't run with your suggested workflow.

Did an alternative that didn't require a csv results file:

```
            fingerprints = pd.read_csv(fingerprint_path).drop(["smiles"], axis=1)
            self.assertAlmostEqual(np.sum(fingerprints.to_numpy()), expected_score, delta=DELTA*expected_score)
```
Ran testing with this aggregation strategy and including an expected score parallel to how the other tests run. Passed 40 times in a row, even with me reducing the DELTA down to 0.0005. This appears to be repeatable on my system. If you're good with using this as the aggregation/test strategy, give it a try on your system and see if it's repeatable there too.

Reply:
Oops, forgot to push the true_fingerprint.csv through the gitignore filters. 

I like your alternative and it seems to be working repeatedly on my system too. What numbers are you using for the expected scores? I'm using 2571.193 and 2840.854 for MPN and last_FFN, just wanted to make sure yours are close.

Reply:
I only ran it for the single MPN, but my number did match yours.